ml-agents/mlagents/trainers/components/reward_signals/gail/model.py

```diff
@@ -0,0 +1,261 @@
+from typing import Tuple, List
+
+import tensorflow as tf
+from mlagents.trainers.models import LearningModel
+
+
+class GAILModel(object):
+    def __init__(
+        self,
+        policy_model: LearningModel,
+        h_size: int = 128,
+        learning_rate: float = 3e-4,
+        encoding_size: int = 64,
+        use_actions: bool = False,
+    ):
+        """
+        The initializer for the GAIL reward generator.
+        https://arxiv.org/abs/1606.03476
+        :param policy_model: The policy of the learning algorithm
+        :param h_size: Size of the hidden layer for the discriminator
+        :param learning_rate: The learning Rate for the discriminator
+        :param encoding_size: The encoding size for the encoder
+        :param use_actions: Whether or not to use actions to discriminate
+        """
+        self.h_size = h_size
+        self.z_size = 128
+        self.alpha = 0.0005
+        self.mutual_information = 0.5
+        self.policy_model = policy_model
+        self.encoding_size = encoding_size
+        self.use_vail = True
+        self.use_actions = use_actions  # True # Not using actions
+        self.make_beta()
+        self.make_inputs()
+        self.create_network()
+        self.create_loss(learning_rate)
+
+    def make_beta(self) -> None:
+        """
+        Creates the beta parameter and its updater for GAIL
+        """
+        self.beta = tf.get_variable(
+            "gail_beta",
+            [],
+            trainable=False,
+            dtype=tf.float32,
+            initializer=tf.ones_initializer(),
+        )
+        self.kl_div_input = tf.placeholder(shape=[], dtype=tf.float32)
+        new_beta = tf.maximum(
+            self.beta + self.alpha * (self.kl_div_input - self.mutual_information), 1e-7
+        )
+        self.update_beta = tf.assign(self.beta, new_beta)
+
+    def make_inputs(self) -> None:
+        """
+        Creates the input layers for the discriminator
+        """
+        self.done_expert = tf.placeholder(shape=[None, 1], dtype=tf.float32)
+        self.done_policy = tf.placeholder(shape=[None, 1], dtype=tf.float32)
+
+        if self.policy_model.brain.vector_action_space_type == "continuous":
```

Comment:
This setup is the same as in BCModel.make_inputs, right? Can it be extracted to a utility method?

Reply:
I agree. Would it be better to put a `utils.py` in the `components` directory, or should it be in one of the two classes and referred from the other class (this doesn't seem too clean)?

Reply:
Not sure, utils.py seems a bit better than enforcing a bad class hierarchy.

Reply:
I'm also a bit reluctant to create a `utils.py` with just one function in it - perhaps we should wait until there are more common functions before refactoring this?